File: js/SHA1/SHA1.js

// SHA1.js
// Ported from c/SHA1/SHA1.c

class SHA1 {
    constructor() {
        this.reset();
    }

    reset() {
        this.state = new Uint32Array(5);
        this.state[0] = 0x67452301;
        this.state[1] = 0xEFCDAB89;
        this.state[2] = 0x98BADCFE;
        this.state[3] = 0x10325476;
        this.state[4] = 0xC3D2E1F0;

        this.count = [0, 0]; // count[0] low, count[1] high (bits)
        this.buffer = new Uint8Array(64);
    }

    static rol(value, bits) {
        return ((value << bits) | (value >>> (32 - bits))) >>> 0;
    }

    static get_uint32_be(b, i) {
        return ((b[i] << 24) | (b[i + 1] << 16) | (b[i + 2] << 8) | b[i + 3]) >>> 0;
    }

    static put_uint32_be(n, b, i) {
        b[i] = (n >>> 24) & 0xFF;
        b[i + 1] = (n >>> 16) & 0xFF;
        b[i + 2] = (n >>> 8) & 0xFF;
        b[i + 3] = n & 0xFF;
    }

    transform(buffer) {
        let a, b, c, d, e;
        const w = new Uint32Array(80);
        let i;

        for (i = 0; i < 16; i++) {
            w[i] = SHA1.get_uint32_be(buffer, i * 4);
        }
        for (i = 16; i < 80; i++) {
            w[i] = SHA1.rol(w[i - 3] ^ w[i - 8] ^ w[i - 14] ^ w[i - 16], 1);
        }

        a = this.state[0];
        b = this.state[1];
        c = this.state[2];
        d = this.state[3];
        e = this.state[4];

        for (i = 0; i < 20; i++) {
            const temp = (SHA1.rol(a, 5) + ((b & c) | (~b & d)) + e + 0x5A827999 + w[i]) >>> 0;
            e = d;
            d = c;
            c = SHA1.rol(b, 30);
            b = a;
            a = temp;
        }

        for (i = 20; i < 40; i++) {
            const temp = (SHA1.rol(a, 5) + (b ^ c ^ d) + e + 0x6ED9EBA1 + w[i]) >>> 0;
            e = d;
            d = c;
            c = SHA1.rol(b, 30);
            b = a;
            a = temp;
        }

        for (i = 40; i < 60; i++) {
            const temp = (SHA1.rol(a, 5) + ((b & c) | (b & d) | (c & d)) + e + 0x8F1BBCDC + w[i]) >>> 0;
            e = d;
            d = c;
            c = SHA1.rol(b, 30);
            b = a;
            a = temp;
        }

        for (i = 60; i < 80; i++) {
            const temp = (SHA1.rol(a, 5) + (b ^ c ^ d) + e + 0xCA62C1D6 + w[i]) >>> 0;
            e = d;
            d = c;
            c = SHA1.rol(b, 30);
            b = a;
            a = temp;
        }

        this.state[0] = (this.state[0] + a) >>> 0;
        this.state[1] = (this.state[1] + b) >>> 0;
        this.state[2] = (this.state[2] + c) >>> 0;
        this.state[3] = (this.state[3] + d) >>> 0;
        this.state[4] = (this.state[4] + e) >>> 0;
    }

    update(data) {
        let input;
        if (typeof data === 'string') {
            input = new TextEncoder().encode(data);
        } else {
            input = new Uint8Array(data);
        }
        
        const len = input.length;
        let j = (this.count[0] >>> 3) & 63;
        
        // Update bit count
        const old_count0 = this.count[0];
        this.count[0] = (this.count[0] + (len << 3)) >>> 0;
        if (this.count[0] < old_count0) {
            this.count[1]++;
        }
        this.count[1] = (this.count[1] + (len >>> 29)) >>> 0;

        let i = 0;
        if ((j + len) > 63) {
            i = 64 - j;
            this.buffer.set(input.subarray(0, i), j);
            this.transform(this.buffer);
            
            for (; i + 63 < len; i += 64) {
                this.transform(input.subarray(i, i + 64));
            }
            j = 0;
        }
        
        this.buffer.set(input.subarray(i, len), j);
    }

    final() {
        const digest = new Uint8Array(20);
        const finalcount = new Uint8Array(8);

        SHA1.put_uint32_be(this.count[1], finalcount, 0);
        SHA1.put_uint32_be(this.count[0], finalcount, 4);

        let i = (this.count[0] >>> 3) & 63;

        this.buffer[i++] = 0x80;

        if (i > 56) {
            while (i < 64) {
                this.buffer[i++] = 0;
            }
            this.transform(this.buffer);
            i = 0;
        }

        while (i < 56) {
            this.buffer[i++] = 0;
        }

        this.buffer.set(finalcount, 56);
        this.transform(this.buffer);

        for (i = 0; i < 5; i++) {
            SHA1.put_uint32_be(this.state[i], digest, i * 4);
        }

        return digest;
    }
}

module.exports = SHA1;

File: c/SHA1/SHA1.c

#include "SHA1.h"
#include <string.h>

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

// Helper to read 32-bit big-endian integer
static uint32_t get_uint32_be(const uint8_t *b, int i)
{
    return ((uint32_t)b[i] << 24) | ((uint32_t)b[i + 1] << 16) |
           ((uint32_t)b[i + 2] << 8) | ((uint32_t)b[i + 3]);
}

// Helper to write 32-bit big-endian integer
static void put_uint32_be(uint32_t n, uint8_t *b, int i)
{
    b[i] = (uint8_t)(n >> 24);
    b[i + 1] = (uint8_t)(n >> 16);
    b[i + 2] = (uint8_t)(n >> 8);
    b[i + 3] = (uint8_t)n;
}

void SHA1Init(SHA1_CTX *context)
{
    context->state[0] = 0x67452301;
    context->state[1] = 0xEFCDAB89;
    context->state[2] = 0x98BADCFE;
    context->state[3] = 0x10325476;
    context->state[4] = 0xC3D2E1F0;
    context->count[0] = context->count[1] = 0;
}

void SHA1Transform(uint32_t state[5], const uint8_t buffer[64])
{
    uint32_t a, b, c, d, e;
    uint32_t w[80];
    int i;

    for (i = 0; i < 16; i++) {
        w[i] = get_uint32_be(buffer, i * 4);
    }
    for (i = 16; i < 80; i++) {
        w[i] = rol(w[i-3] ^ w[i-8] ^ w[i-14] ^ w[i-16], 1);
    }

    a = state[0];
    b = state[1];
    c = state[2];
    d = state[3];
    e = state[4];

    for (i = 0; i < 20; i++) {
        uint32_t temp = rol(a, 5) + ((b & c) | (~b & d)) + e + 0x5A827999 + w[i];
        e = d;
        d = c;
        c = rol(b, 30);
        b = a;
        a = temp;
    }

    for (i = 20; i < 40; i++) {
        uint32_t temp = rol(a, 5) + (b ^ c ^ d) + e + 0x6ED9EBA1 + w[i];
        e = d;
        d = c;
        c = rol(b, 30);
        b = a;
        a = temp;
    }

    for (i = 40; i < 60; i++) {
        uint32_t temp = rol(a, 5) + ((b & c) | (b & d) | (c & d)) + e + 0x8F1BBCDC + w[i];
        e = d;
        d = c;
        c = rol(b, 30);
        b = a;
        a = temp;
    }

    for (i = 60; i < 80; i++) {
        uint32_t temp = rol(a, 5) + (b ^ c ^ d) + e + 0xCA62C1D6 + w[i];
        e = d;
        d = c;
        c = rol(b, 30);
        b = a;
        a = temp;
    }

    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;
    state[4] += e;
}

void SHA1Update(SHA1_CTX *context, const uint8_t *data, uint32_t len)
{
    uint32_t i, j;

    j = (context->count[0] >> 3) & 63;
    if ((context->count[0] += len << 3) < (len << 3))
        context->count[1]++;
    context->count[1] += (len >> 29);

    if ((j + len) > 63) {
        memcpy(&context->buffer[j], data, (i = 64 - j));
        SHA1Transform(context->state, context->buffer);
        for (; i + 63 < len; i += 64) {
            SHA1Transform(context->state, &data[i]);
        }
        j = 0;
    }
    else i = 0;
    memcpy(&context->buffer[j], &data[i], len - i);
}

void SHA1Final(uint8_t digest[20], SHA1_CTX *context)
{
    uint32_t i;
    uint8_t finalcount[8];
    
    put_uint32_be(context->count[1], finalcount, 0);
    put_uint32_be(context->count[0], finalcount, 4);

    i = (context->count[0] >> 3) & 63;
    
    context->buffer[i++] = 0x80;
    
    if (i > 56) {
        while (i < 64) {
            context->buffer[i++] = 0;
        }
        SHA1Transform(context->state, context->buffer);
        i = 0;
    }
    
    while (i < 56) {
        context->buffer[i++] = 0;
    }
    
    memcpy(&context->buffer[56], finalcount, 8);
    SHA1Transform(context->state, context->buffer);
    
    for (i = 0; i < 5; i++) {
        put_uint32_be(context->state[i], digest, i * 4);
    }
}

void SHA1(const uint8_t *data, uint32_t len, uint8_t digest[20])
{
    SHA1_CTX ctx;
    SHA1Init(&ctx);
    SHA1Update(&ctx, data, len);
    SHA1Final(digest, &ctx);
}
